src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp

```diff
@@ -0,0 +1,127 @@
+ */
+#ifndef MLPACK_METHODS_XGBOOST_LOSS_FUNCTIONS_SSE_LOSS_HPP
+#define MLPACK_METHODS_XGBOOST_LOSS_FUNCTIONS_SSE_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ensemble {
+
+/**
+ * The SSE (Sum of Squared Errors) loss is a loss function to measure the
+ * quality of prediction of response values present in the node of each
+ * xgboost tree. It is also a good measure to compare the spread of two
+ * distributions. We will try to minimize this value while training.
+ *
+ * Loss = 1 / 2 * (Observed - Predicted)^2
+ */
+class SSELoss
+{
+ public:
+  /**
+   * Returns the initial predition for gradient boosting.
+   */
+  template<typename VecType>
+  typename VecType::elem_type InitialPrediction(const VecType& values)
+  {
+    return arma::accu(values) / (typename VecType::elem_type) values.n_elem;
+  }
+
+  /**
+   * Returns the first order gradient of the loss function with respect to the
+   * values.
+   *
+   * This is primarily used in calculating the residuals and split gain for the
+   * gradient boosted trees.
+   *
+   * @tparam T The type of input data. This can be both a vector or a scalar.
+   * @param observed The true observed values.
+   * @param values The values with respect to which the gradient will be
+   *     calculated.
+   */
+  template<typename T>
+  T Gradients(const T& observed, const T& values)
+  {
+    return - (observed - values);
+  }
+
+  /**
+   * Returns the second order gradient of the loss function with respect to the
+   * values. This is used only for scalars.
+   */
+  template<typename T>
+  T Hessians(const T& /* observed */, const T& /* values */)
+  {
+    return (T) 1;
+  }
+
+  /**
+   * Returns the second order gradient of the loss function with respect to the
+   * values. This is used only for vectors.
+   */
+  template<typename VecType,
+           class = std::enable_if_t<arma::is_Col<VecType>::value ||
+               arma::is_Row<VecType>::value>>
+  VecType Hessians(const VecType& /* observed */, const VecType& values)
+  {
+    VecType h(values.n_elem, arma::fill::ones);
+    return h;
+  }
+
+  /**
+   * Returns the pseudo residuals of the predictions.
+   * This is equal to the negative gradient of the loss function with respect
+   * to the predicted values f.
+   *
+   * @param observed The true observed values.
+   * @param f The prediction at the current step of boosting.
+   */
+  template<typename VecType>
+  VecType Residuals(const VecType& observed, const VecType& f)
+  {
+    return - Gradients(observed, f);
+  }
+
+  /**
+   * Returns the output value for the leaf in the tree.
+   */
+  template<typename VecType>
+  typename VecType::elem_type
+  OutputValue(const VecType& gradients, const VecType& hessians,
+              const double lambda)
+  {
+    return - arma::accu(gradients) / (arma::accu(hessians) + lambda);
+  }
+
+  /**
+   * Calculates the similarity score for evaluating the splits.
```

Comment:
If I understand the algorithm correctly, I guess this is really the `FitnessFunction` that we will use to evaluate splits.  Maybe we should implement the same API, such that we can use this class as the `FitnessFunction`?  I could see this working if, e.g., we passed a *two*-dimensional matrix (instead of a one-dimensional vector) to the fitness function for the responses/labels, where the two dimensions were the gradients and hessians, respectively.  I believe that this could work since the fitness functions' `Evaluate()` method should be templatized and able to take a matrix instead of a vector.  (If that is not true, it seems like it could be a simple refactoring to make that work.)

Reply:
Yeah! This could work. The Evaluate() method is already templatized so it should work without any refactoring. I will make the API consistent with the existing FitnessFunction classes.

Reply:
@rcurtin I just now realized that we also need regularization parameters lambda and alpha for gain computation in XGBoost. But those parameters are not needed in the normal decision tree. So will it be okay to add these parameters to existing FitnessFunctions for making them consistent?